### ISSUE TITLE
Using `nn` without `-Yexplicit-nulls` Crashes in RC1 but not M3

### ISSUE THREAD

**Issue by AugustNagro** [lampepfl/dotty#11631]
## Compiler version

3.0.0-RC1

(No crash in 3.0.0-M3)

## Minimized code

```Scala
trait MyTrait:
  def a(): String = ""

class MyClass:
  var myTrait: MyTrait|Null = null

  def printA(): Unit = println(myTrait.nn.a())

@main def runTest(): Unit =
  val mt = new MyTrait:
    override def a(): String = "hello world"

  val mc = MyClass()
  mc.myTrait = mt
  mc.printA()
```

## Output

```
undefined: {
  val x$proxy1: (MyTrait | Null) = this.myTrait
  (scala.runtime.Scala3RunTime.nn[MyTrait](x$proxy1):x$proxy1.type & MyTrait)
}.a # -1: TermRef(OrType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),trait MyTrait),TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Null)),a) at inlining
```

Enabling explicit nulls fixes this, but I think the program should still compile. I ran into this problem when making an (encapsulated) Scala.js [Web Component](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements). Web Components may only have 0-arg constructors.

**Comment by AugustNagro** [comment]
Oh, I just read that `Don't use .nn on mutable variables directly, because it may introduce an unknown type into the type of the variable.`

Is that the case here? And would the solution be replacing `nn` with `instanceof[MyTrait]`?

**Comment by odersky** [comment]
Here's a minimization without `nn` or `Yexplicit-nulls`:
```scala
trait MyTrait:
  def a(): String = ""

class Nul

extension [T](x: T | Nul) inline def nnn: x.type & T = ???

class MyClass:
  var myTrait: MyTrait|Null = null

  def printA(): Unit = println(myTrait.nnn.a())

@main def runTest(): Unit =
  val mt = new MyTrait:
    override def a(): String = "hello world"

  val mc = MyClass()
  mc.myTrait = mt
  mc.printA()
```
It happens at inliner phase. Seems to be a consequence of moving inlining to its own phase.

**Comment by odersky** [comment]
In fact, things seem to work if we replace the definition of `myTrait` above with
```scala
var myTrait: MyTrait|Nul = ???
```
So it does seem to be related to `Null` handling in some way, after all. 

Maybe a stopgap measure is to disallow `nn` unless `-YexplicitNulls` is set?

**Comment by AugustNagro** [comment]
Sounds good to me. Seems like a rare / low priority bug. I only encountered
by somehow forgetting to set explicit nulls.

Thanks for looking into it!

**Comment by OlivierBlanvillain** [comment]
Closed in #13604
